File: js/map.js
import { MAP_OFFSET_X, MAP_OFFSET_Y, TILE_SIZE, LEVEL_LAYOUT } from './constants.js';

export class GameMap {
    constructor() {
        this.walls = [];
        this.destructibles = [];
        
        // Carga de imágenes de texturas del escenario
        this.wallImg = new Image();
        this.wallImg.src = 'assets/wall.png';
        
        this.destImg = new Image();
        this.destImg.src = 'assets/destructible.png';

        this.initMap();
    }

    // Inicializa la lectura del mapa mediante la matriz de constantes
    initMap() {
        LEVEL_LAYOUT.forEach((row, r) => {
            for (let c = 0; c < row.length; c++) {
                const char = row[c];
                const x = c * TILE_SIZE + MAP_OFFSET_X;
                const y = r * TILE_SIZE + MAP_OFFSET_Y;
                const rect = { x, y, width: TILE_SIZE, height: TILE_SIZE };

                if (char === 'W') this.walls.push(rect);
                if (char === 'D') this.destructibles.push(rect);
            }
        });
    }

    // Filtra y elimina bloques destructibles alcanzados por una explosión
    removeBlocksInArea(explosionRect) {
        this.destructibles = this.destructibles.filter(block => {
            return !(block.x < explosionRect.x + explosionRect.width &&
                     block.x + block.width > explosionRect.x &&
                     block.y < explosionRect.y + explosionRect.height &&
                     block.y + block.height > explosionRect.y);
        });
    }

    // Dibuja los muros y bloques destructibles en pantalla
    draw(ctx) {
        this.walls.forEach(w => ctx.drawImage(this.wallImg, w.x, w.y, TILE_SIZE, TILE_SIZE));
        this.destructibles.forEach(d => ctx.drawImage(this.destImg, d.x, d.y, TILE_SIZE, TILE_SIZE));
    }
}

File: Gatito Enojado - Python + Pygame/src/map.py
"""
Módulo de Mapa
Construye y renderiza el escenario del juego a partir de una matriz de texto (layout).
"""

import os
import sys

import pygame

from src.constants import MAP_OFFSET_X, MAP_OFFSET_Y, TILE_SIZE


def resource_path(relative_path):
    """Obtiene la ruta absoluta segura para los archivos de texturas."""
    try:
        base_path = sys._MEIPASS
    except Exception:  # noqa: BLE001
        base_path = os.path.abspath(".")
    return os.path.join(base_path, relative_path)

class Map:
    def __init__(self, layout):
        # Carga de texturas para muros fijos y bloques destruibles
        self.wall_img = pygame.image.load(resource_path('assets/wall.png')).convert_alpha()
        self.wall_img = pygame.transform.scale(self.wall_img, (TILE_SIZE, TILE_SIZE))
        
        self.dest_img = pygame.image.load(resource_path('assets/destructible.png')).convert_alpha()
        self.dest_img = pygame.transform.scale(self.dest_img, (TILE_SIZE, TILE_SIZE))
        
        self.walls = []
        self.destructibles = []
        
        # Procesamiento del diseño del mapa basado en caracteres
        for r, row in enumerate(layout):
            for c, char in enumerate(row):
                x = c * TILE_SIZE + MAP_OFFSET_X
                y = r * TILE_SIZE + MAP_OFFSET_Y
                if char == 'W': 
                    self.walls.append(pygame.Rect(x, y, TILE_SIZE, TILE_SIZE))
                if char == 'D': 
                    self.destructibles.append(pygame.Rect(x, y, TILE_SIZE, TILE_SIZE))

    def remove_blocks_in_area(self, rect):
        """Elimina los bloques destructibles que entren en contacto con un área específica (explosiones)."""
        self.destructibles = [d for d in self.destructibles if not d.colliderect(rect)]

    def draw(self, screen):
        """Renderiza todos los elementos estáticos del mapa en la pantalla."""
        for w in self.walls: 
            screen.blit(self.wall_img, (w.x, w.y))
        for d in self.destructibles: 
            screen.blit(self.dest_img, (d.x, d.y))